core/src/Kokkos_Complex.hpp

```diff
@@ -423,6 +440,43 @@ class
 #endif  // KOKKOS_ENABLE_DEPRECATED_CODE_4
 };
 
+}  // namespace Kokkos
+
+// Tuple protocol for complex based on https://wg21.link/p2819r2
+template <typename RealType>
+struct std::tuple_size<Kokkos::complex<RealType>>
+    : std::integral_constant<size_t, 2> {};
+
+template <size_t I, typename RealType>
+struct std::tuple_element<I, Kokkos::complex<RealType>> {
+  static_assert(I < 2);
+  using type = RealType;
+};
+
+namespace Kokkos {
+
+template <size_t I, typename RealType>
+KOKKOS_FUNCTION constexpr RealType& get(complex<RealType>& z) noexcept {
+  return Impl::complex_get<I>(z);
```

Comment:
Why did you chose to introduce `Impl::complex_get` instead of having the various `get` be declared as hidden friends?

Reply:
It doesn't work as hidden friends.  That was the mistake LEWG made with P2819R1 (and was corrected by LWG for P2819R2).

Reply:
I had briefly looked in libcxx and they did friends but public

Reply:
One can also do SBs via member `get` functions (harder to use in general, but that doesn’t matter if the intent is only to have these function to support SBs).

I initially had that approach working, but abandoned it as I didn’t want the interface to depart from that of `std::complex` any more than necessary.  Should I resurrect it?

Reply:
> It doesn't quite work? @crtrott, you were present for the LWG discussion of P2819R1 in Kona (I was not) when LWG told LEWG they were making it non-hidden. Thoughts?
> 
> As for "public", are you just saying you want me to move where the friend declaration is in the class?

I am suggesting you define `get` in the body of the class as a public friend function and you inline the definition and drop the `Impl::complex_get` helper.

Reply:
@dalg24 So you are basically saying that you want me to implement [P2819R1](https://wg21.link/P2819R1), not [P2819R2](https://wg21.link/P2819R2) (and I assume remove the comment about this being approved for C++26, because P2819R1 was rejected by LWG).

I don't see the justification for going against what was ratified for C++26.

Reply:
I expect if you forward declare the get function and implement it as a friend that will follow the spec from R2 that is in the current draft. I slightly prefer doing it that way but won't fight if you disagree,

Reply:
I'll do that (I thought you were proposing an interface change, not an implementation change).

I think I'll also change `Impl::complex_get` to a `private static` function in `complex` so it won't need friendship.  It was in `Impl` because it is somewhat independent of `T`, as its main concern is the correct reference type, but that required it being a `friend` because `.real()` and `.imag()` don't always return references.

Reply:
I don't think you can forward declare `friend` functions for a templated class, which means they are still hidden friends.

If I use `__PRETTY_FUNCTION__` on:

```c++
template <size_t I, typename RealType>
constexpr RealType& get(complex<RealType>& z) noexcept { /* ... */ }
```
It returns

`RealType &Kokkos::get(complex<RealType> &) [I = 0UL, RealType = double]`

If I use `__PRETTY_FUNCTION__` on:

```c++
  template <size_t I>
friend constexpr RealType& get(complex& z) noexcept { /* ... */ }
```

it returns

`RealType &Kokkos::get(complex<RealType> &) [I = 0UL]`

If I put both definitions in, structured bindings calls the `friend` function.

I'll give it some thought over the weekend, but I'm leaning towards reverting it, as this involves some dark corners of the language...

Reply:
Changed.  Made `get` a direct friend declaration in the two cases it is needed.